pkg/cloudprovider/provider/aws/provider.go

```diff
@@ -170,78 +207,119 @@ type amiFilter struct {
-			},
-			{
-				Name:   aws.String("root-device-type"),
-				Values: aws.StringSlice([]string{"ebs"}),
-			},
-			{
-				Name:   aws.String("architecture"),
-				Values: aws.StringSlice([]string{"x86_64"}),
+		cacheKey := fmt.Sprintf("ami-id-%s-%s-%s", region, os, arch)
+		amiID, found := cache.Get(cacheKey)
+		if found {
+			klog.V(3).Info("found AMI-ID in cache!")
+			return amiID.(string), nil
+		}
+
+		describeImagesInput := &ec2.DescribeImagesInput{
+			Owners: aws.StringSlice([]string{filter.owner}),
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("description"),
+					Values: aws.StringSlice([]string{filter.description}),
+				},
+				{
+					Name:   aws.String("virtualization-type"),
+					Values: aws.StringSlice([]string{"hvm"}),
+				},
+				{
+					Name:   aws.String("root-device-type"),
+					Values: aws.StringSlice([]string{"ebs"}),
+				},
+				{
+					Name:   aws.String("architecture"),
+					Values: aws.StringSlice([]string{string(arch)}),
+				},
 			},
-		},
-	}
+		}
 
-	if filter.productCode != "" {
-		describeImagesInput.Filters = append(describeImagesInput.Filters, &ec2.Filter{
-			Name:   aws.String("product-code"),
-			Values: aws.StringSlice([]string{filter.productCode}),
-		})
+		if filter.productCode != "" {
+			describeImagesInput.Filters = append(describeImagesInput.Filters, &ec2.Filter{
+				Name:   aws.String("product-code"),
+				Values: aws.StringSlice([]string{filter.productCode}),
+			})
+		}
+
+		imagesOut, err := client.DescribeImages(describeImagesInput)
+		if err != nil {
+			return "", err
+		}
+
+		if len(imagesOut.Images) == 0 {
+			klog.V(3).Infof("could not find image for '%s' with CPU architecture '%s'", os, arch)
+			continue
+		}
+
+		if os == providerconfigtypes.OperatingSystemRHEL {
+			imagesOut.Images, err = filterSupportedRHELImages(imagesOut.Images)
+			if err != nil {
+				return "", err
+			}
+		}
+
+		image := imagesOut.Images[0]
+		for _, v := range imagesOut.Images {
+			itime, _ := time.Parse(time.RFC3339, *image.CreationDate)
+			vtime, _ := time.Parse(time.RFC3339, *v.CreationDate)
+			if vtime.After(itime) {
+				image = v
+			}
+		}
+
+		cache.SetDefault(cacheKey, *image.ImageId)
+		return *image.ImageId, nil
 	}
 
-	imagesOut, err := client.DescribeImages(describeImagesInput)
+	return "", fmt.Errorf("could not find suitable image for '%s' with any supported CPU architecture", os)
+}
+
+func getCPUArchitectures(client *ec2.EC2, instanceType string) ([]awstypes.CPUArchitecture, error) {
+	// read the instance type to know which cpu architecture is needed in the AMI
+	instanceTypes, err := client.DescribeInstanceTypes(&ec2.DescribeInstanceTypesInput{
+		InstanceTypes: []*string{aws.String(instanceType)},
+	})
+
 	if err != nil {
-		return "", err
+		return []awstypes.CPUArchitecture{}, err
 	}
 
-	if len(imagesOut.Images) == 0 {
-		return "", fmt.Errorf("could not find Image for '%s'", os)
+	if len(instanceTypes.InstanceTypes) != 1 {
```

Comment:
since we (rightfully so) check for exactly 1 instance type, it will have exactly 1 CPU architecture, so I suggest to return only 1 `awstypes.CPUArchitecture`

Reply:
Changed things so only one awstypes.CPUArchitecture is used now :)